### DISCUSSION TITLE
Verwendung als Library: Wie erstellt man Container und was ist value

### DISCUSSION THREAD

**Discussion by wireva** [spali/go-rscp#41] [Q&A]
Hi,
Ich versuch go-rscp als library in go zu verwenden.

Sieht aktuell so aus:
```go
func main() {
	fmt.Println("hi")

	config := rscp.ClientConfig{
		Address:  "192.168.178.21",
		Port:     5033,
		Username: "local.user",
		Password: "123456",
		Key:      "123456",
	}
	client, err := rscp.NewClient(config)
	if err != nil {
		panic(err)
	}
	defer client.Disconnect()

	var thenumber uint8 = 0
	// message := rscp.NewMessage(rscp.INFO_REQ_SERIAL_NUMBER, nil)
	// rscp.PVI_REQ_MAX_TEMPERATURE
	message := rscp.NewMessage(rscp.PVI_REQ_MAX_TEMPERATURE, &thenumber)

	response, err := client.Send(*message)
	if err != nil {
		panic(err)
	}

	fmt.Println(response.Value)
}
```

Hier bekomme ich folgenden Fehler:
panic: message at index 0: expected *uint8 got *uint8 : value does not match data type

was irgendwie wenig sinn macht für mich... 
Was muss ich als value bei NewMessage angeben? Bei manchen Anfragen funktioniert nil.

Und dann gibt es tags die ich nicht abfragen kann weil sie einen "container" brauchen (viele Batterieanfragen zb). Wie erstelle ich den?

**Comment by spali** [comment]
Bin gerade unterwegs, darum kann ich nicht genauer in den Code schauen.
Schau mal `CreateRequest` an. Ist gedacht um requests zu erstellen.
Da ist der Kommentar auch ein wenig genauer, auch im Zusammenhang mit Container.

**Reply by wireva** [reply]
danke, das macht schonmal deutlich mehr sinn. Wo findet man die Dokumentation von E3DC zu dem Protokoll? Weil ich hab keine Ahnung was ich da für Werte hinsenden muss und was was ist. Zum Beispiel versuche ich die Temperatur vom Wechselrichter zu bekommen. hab es versucht mit PVI_REQ_MIN_TEMPERATURE, aber das will als value einen *uint8 haben. Ich denke mal das ist die ID vom Wechselrichter? Aber wenn ich da irgendeine Zahl als *uint8 einsetze bekomme ich immer die Fehlermeldung `expected *uint8 got *uint8 : value does not match data type`

**Reply by spali** [reply]
Sollte online bei E3DC downloadbar sein für deine E3DC. Ist vermutlich als Programmierer Beispiel benannt. Darin war bei mir eine Excel Datei mit allen TAG's und deren Datentypen. Kann leichte Unterschiede zu den Tags in der library haben. Denke aber das gröbste sollte das gleiche sein.
Container musst du nicht explizit erstellen. Das ist ein Datentyp der in der library einfach einem array von `Message`s entspricht. Die json Beispiele im Wiki machen das ev. besser deutlich wie die Messages verschachtelt werden.
Du kannst auch mal mein code vom binary nehmen als Beispiel und Debugging einstellen oder sogar selber was ausgeben an der Stelle vo du etwas nicht verstehst, was die library erwartet. 
Sorry aber meine Doku ist auch nicht viel besser als die spährlichen Infos von E3DC 😬
